packages/graphile-build-pg/src/QueryBuilder.js

```diff
@@ -868,11 +886,16 @@ order by (row_number() over (partition by 1)) desc`; /* We don't need to factor
       this.locks[type] = isDev ? new Error("Initally locked here").stack : true;
       this.compiledData[type] = data;
     } else if (type === "orderBy") {
-      this.compiledData[type] = this.data[type].map(([a, b, c]) => [
-        callIfNecessary(a, context),
-        b,
-        c,
-      ]);
+      // Add distinct to existing order by to avoid `SELECT DISTINCT ON expressions must match initial ORDER BY expressions`
+      // Convert distinct fields to order by applying ascending and using existing order by for matching fields
+      const distinctAsOrderBy = this.data["distinctOn"].map(
+        d => this.data[type].find(([a]) => isEqual(a, d)) ?? [d, true, null]
+      );
+
+      // Create a union of the distinct and order by fields
+      this.compiledData[type] = [
+        ...new Set([...distinctAsOrderBy, ...this.data[type]]),
+      ].map(([a, b, c]) => [callIfNecessary(a, context), b, c]);
```

Comment:
I don't want to incur the performance costs of the `distinctOn` logic when distinctOn is not used (which will be in the _vast_ majority of schemas). Therefore, please keep the original code, and wrap it with `if (this.data.distinctOn.length === 0) { /* old code */ } else { /* new code */ }`